packages/devtools_app/lib/src/performance/performance_model.dart

```diff
@@ -427,14 +427,30 @@ class FlutterFrame {
       pipelineItemTime.start?.inMicroseconds != null &&
       pipelineItemTime.end?.inMicroseconds != null;
 
-  int get uiDuration => uiEventFlow?.time?.duration?.inMicroseconds;
-
-  double get uiDurationMs => uiDuration != null ? uiDuration / 1000 : null;
+  Duration get uiDuration => uiEventFlow?.time?.duration;
+
+  Duration get rasterDuration => rasterEventFlow?.time?.duration;
+
+  Duration get shaderDuration {
+    if (_shaderTime != null) return _shaderTime;
+    int sumShaderMicros = 0;
+    breadthFirstTraversal<TimelineEvent>(
+      rasterEventFlow,
+      action: (TimelineEvent event) {
+        // Shader events with a shader event parent should not be included in the
+        // sum because the parent event will encompass the time of [event].
+        if (event.isShaderEvent &&
```

Comment:
to make this less fragile: we could union all intervals of time with shader events. Add a TODO to add a helper class for performing efficient time interval union computation. Basically you should just be able to call addInterval repeatedly on the class and then query for the total interval without relying on  a specific structure of shader events. This current code could be broken if there ever happen to be non-shader events interleaved with shader events in which case it would report overly large ranges.

Reply:
added todo